src/+otp/+utils/+solvers/dae43.m

```diff
@@ -0,0 +1,249 @@
+    end
+
+    H = Mfull - h*Jfull;
+    if usesparse
+        [L, U, P, Q, D] = lu(H);
+    else
+        [L, U] = lu(H);
+    end
+
+    while kappa*(etak*nnp) >= ntol && its < nmaxits
+        % build g and Jacobians
+        for stage = 1:stagenum
+            si = ((stage - 1)*n + 1):(stage*n);
+            staget = tc + c(stage)*h;
+            ycs = yc + h*reshape(newtonk, n, [])*(A(stage, :).');
+            sc(si) = abstol + reltol*abs(ycs);
+            Mc = Mfull(si, si);
+            gfull(si) =  Mc*newtonk(si) - f(staget, ycs);
+        end
+
+        if usesparse
+            npnew = Q*(U\(L\(P*(D\gfull))));
+        else
+            npnew = U\(L\gfull);
+        end
+
+        if its > 1
+            nnpnew = sqrt(mean((npnew./sc).^2));
+
+            thetak = nnpnew/nnp;
+
+            if thetak > 1.25 || isnan(thetak) || isinf(thetak)
+                
+                % If theta is large, we recompute the Jacobian and try again
+                if ~brejectstage
+                    brejectstage = true;
+
+
+                    for stage = 1:stagenum
+                        si = ((stage - 1)*n + 1):(stage*n);
+                        staget = tc + c(stage)*h;
+                        ycs = yc + h*reshape(newtonk, n, [])*(A(stage, :).');
+                        Jc = J(staget, ycs);
+                        Jfull(si, :) = kron(A(stage, :), Jc);
+                    end
+                    H = Mfull - h*Jfull;
+                    if usesparse
+                        [L, U, P, Q, D] = lu(H);
+                    else
+                        [L, U] = lu(H);
+                    end
+
+                    continue;
+                end
+
+                % If theta is still large after recomputing the Jacobian, we reject the step and decrease the stepsize
+
+                bnewtonreject = true;
+                break;
+            end
+
+            brejectstage = false;
+
+            etak = thetak/(1 - thetak);
+        end
+
+        newtonk = newtonk - npnew;
+
+        np = npnew;
+        nnp = sqrt(mean((np./sc).^2));
+
+        its = its + 1;
+    end
+
+    if bnewtonreject
+       h = h/2;
+       continue;
+    end
+
+    yhat = yc + h*reshape(newtonk, n, [])*bhat.';
+    ycnew = yc + h*reshape(newtonk, n, [])*b.';
+
+    sc = abstol + max(abs(ycnew), abs(yc))*reltol;
+
+    Mc = M(tc + h , ycnew);
+
+    err = sqrt(mean(((Mc*(ycnew - yhat))./sc).^2));
+
+    fac = 0.9;
+
+    facmin = 0.1;
+    if err > 1 || isnan(err)
+        % Reject timestep
+        facmax = 1;
+    else
+        % Accept time step
+        tc = tc + h;
+
+        yc = ycnew;
+
+        t(step + 1) = tc;
+        y(step + 1, :) = yc.';
+
+
+        step = step + 1;
+
+        facmax = 4;
+    end
+
+    % adjust step-size
+    h = h*min(facmax, max(facmin, fac*(1/err)^1/(orderE + 1)));
+
+end
+
+if nargout < 2
+    sol = struct('x', t, 'y', y.');
+else
+    sol = t;
+end
+
+end
+
+function J = jacapprox(f, t, y)
```

Comment:
This seems like a useful enough to warrant it's own publicly-accessible function.  We can also look into more sophisticated and robust approaches.